### PR TITLE
feat: change handling of new data when internal queue is full

### PR DESCRIPTION
**Pull request by idobes**

### CODE REVIEW

File: era_5g_interface/task_handler_internal_q.py [hunk 1]
```diff
@@ -18,18 +19,39 @@ class TaskHandlerInternalQ(TaskHandler):
     method externally.
     """
 
-    def __init__(self, sid: str, image_queue: Queue, **kw) -> None:
+    def __init__(self, sid: str, image_queue: Queue, if_queue_full="discard_oldest", queue_put_lock=None, **kw) -> None:
```
Comment (Kapim) [inline era_5g_interface/task_handler_internal_q.py:22]: if_queue_full - can you make it an enum or something other than a string?

Comment (Kapim) [reply era_5g_interface/task_handler_internal_q.py:22]: and please rename the image_queue to data_queue

File: era_5g_interface/task_handler_internal_q.py [hunk 2]
```diff
@@ -45,13 +67,40 @@ def store_image(self, metadata: dict, image: np.ndarray) -> None:
                 The format is NetApp-specific.
             image (_type_): The image to be processed.
         """
+
```
Comment (Kapim) [inline era_5g_interface/task_handler_internal_q.py:70]: please rename the method to "store_data" and the field image to data and put there Any as a anotation

**Review by Kapim — CHANGES_REQUESTED** [review]

File: era_5g_interface/task_handler_internal_q.py (hunk 1 above)
Comment (ZdenekM) [reply era_5g_interface/task_handler_internal_q.py:22]: Please add a type annotation to `queue_put_lock`.

File: era_5g_interface/task_handler_internal_q.py [hunk 3]
```diff
@@ -45,13 +67,40 @@ def store_image(self, metadata: dict, image: np.ndarray) -> None:
...
+        # make sure multiple calls of store_image do not interfere with each other
+        self._q_put_lock.acquire()
```
Comment (ZdenekM) [inline era_5g_interface/task_handler_internal_q.py:72]: Why not `with self._q_put_lock:`?

File: era_5g_interface/task_handler_internal_q.py [hunk 4]
```diff
@@ -45,13 +67,40 @@ def store_image(self, metadata: dict, image: np.ndarray) -> None:
...
             self._q.put((metadata, image), block=False)
         except Full:
-            pass
-            # TODO: raise an exception
+            if self.if_queue_full == "discard_new":
+                pass  # do nothing (discard the new data)
+            elif self.if_queue_full == "discard_oldest":
+                # remove the oldest data from queue to make up space for the new one,
+                # but make sure not to discard control commands
+                self._q.mutex.acquire()  # acquire the queue lock to prevent it from mutating during the removal
```
Comment (ZdenekM) [inline era_5g_interface/task_handler_internal_q.py:84]: Isn't one lock (`self._q_put_lock`) enough? Or couldn't we use just the "queue lock"?

File: era_5g_interface/task_handler_internal_q.py [hunk 5]
```diff
@@ -45,13 +67,40 @@ def store_image(self, metadata: dict, image: np.ndarray) -> None:
...
+                        logger.warning(
+                            "Discarding new data instead of the old one to prevent losing a control command in queue."
+                        )
+                        # this situation could also raise an exception, if desired
+                finally:
+                    self._q.mutex.release()
+                if space_available:
+                    self._q.put((metadata, image), block=False)
+            else:
+                raise RuntimeError("Internal queue is full and no data was allowed to be discarded.")
```
Comment (ZdenekM) [inline era_5g_interface/task_handler_internal_q.py:101]: Why `RuntimeError`? I would rather use some custom exception.

**Review by ZdenekM — CHANGES_REQUESTED** [review]

File: era_5g_interface/task_handler_internal_q.py (hunk 1 above)
Comment (idobes) [reply era_5g_interface/task_handler_internal_q.py]: done

File: era_5g_interface/task_handler_internal_q.py (hunk 2 above)
Comment (idobes) [reply era_5g_interface/task_handler_internal_q.py:101]: done

File: era_5g_interface/task_handler_internal_q.py (hunk 5 above)
Comment (idobes) [reply era_5g_interface/task_handler_internal_q.py]: Changed to QueueDiscardError

File: era_5g_interface/task_handler_internal_q.py (hunk 3 above)
Comment (idobes) [reply era_5g_interface/task_handler_internal_q.py]: Yes, I agree, it makes the code cleaner. Changed.

File: era_5g_interface/task_handler_internal_q.py (hunk 4 above)
Comment (ZdenekM) [reply era_5g_interface/task_handler_internal_q.py]: Sure, we can meet, but if it is essential, it should be fully explained in the source code so we will know the reason even next week or next month :-)

File: era_5g_interface/task_handler_internal_q.py [hunk 6]
```diff
@@ -1,70 +1,161 @@
...
-            pass
-            # TODO: raise an exception
+        with self._q_put_lock:
+            try:
+                self._q.put(data, block=False)
+            except Full:
+                if self.if_queue_full == QueueFullAction.DISCARD_OLDEST:
+                    with self._q.mutex:
+                        type_of_old_data = type(self._q.queue[0])
+                        if type_of_old_data is not ControlCommand:
```
Comment (ZdenekM) [inline era_5g_interface/task_handler_internal_q.py:149]: Please use `if isinstance` as well.

**Comment by ZdenekM** [comment]
@idobes Also, please update the version to `0.6.0` here https://github.com/5G-ERA/era-5g-interface/blob/main/era_5g_interface/BUILD#L12 so it's not necessary to do that in a separate commit. Thanks!

File: era_5g_interface/task_handler_internal_q.py (hunk 6 above)
Comment (idobes) [reply era_5g_interface/task_handler_internal_q.py]: Sorry, I forgot about this one.. Fixed.

**Comment by idobes** [comment]
> @idobes Also, please update the version to `0.6.0` here https://github.com/5G-ERA/era-5g-interface/blob/main/era_5g_interface/BUILD#L12 so it's not necessary to do that in a separate commit. Thanks!
Version updated.

**Review by ZdenekM — APPROVED** [review]

**Review by Kapim — APPROVED** [review]

**Pull request merged**